### ISSUE TITLE
EKF2 replay on SITL logs is not reproducible

### ISSUE THREAD

**Issue by vanurag** [PX4/Firmware#14651]
**Describe the bug**
I am using EKF2 replay to tune some of my parameters. As a starting point, I wanted to see if I can reproduce similar ekf2 output when I replay a log with unchanged parameters. I used jmavsim to perform a short flight session. I then performed ekf2 replay on this log. I am using Tools/ecl_ekf/process_logdata_ekf.py for generating analysis plots for the original log and the replayed log. I notice that they differ quite significantly.

**Log Files and Screenshots**
- [Original log file](https://review.px4.io/plot_app?log=407615c9-9d13-453b-b988-fbcd05d8f637)
- [Replayed log file](https://review.px4.io/plot_app?log=1ef81b28-cd83-4def-ada3-58995a109957)

**EKF analysis plots**
- [Original log analysis](https://github.com/PX4/Firmware/files/4469426/11_22_21.ulg.pdf)
- [Replayed log analysis](https://github.com/PX4/Firmware/files/4469427/11_23_20_replayed.ulg.pdf)

**Comment by dagar** [comment]
@bresch @jkflying @kamilritz FYI

**Comment by vanurag** [comment]
Hey guys, any update on this issue? I saw that a new issue for [unit-testing EKF replay](https://github.com/PX4/Firmware/issues/14695) has been opened. Does that mean the EKF replay in it's current state is not functional?

**Comment by priseborough** [comment]
I've run a check using a log generated using jmavsim built using master at commit https://github.com/PX4/Firmware/commit/326d8efc166331d5b9a0fc96836b68e82142a409 and it looks like there are two issues.

The first, which is easily fixed is that during replay, the logging of the uORB data does not start until after 20 seconds. This can be fixed by reducing the number of microseconds between subscription retries in logger.h, eg:

replace

static constexpr hrt_abstime TRY_SUBSCRIBE_INTERVAL{1000 * 1000};	// interval in microseconds at which we try to subscribe to a topic

with

// interval in microseconds at which we try to subscribe to a topic
#ifdef ORB_USE_PUBLISHER_RULES
static constexpr hrt_abstime TRY_SUBSCRIBE_INTERVAL {1000};
#else
static constexpr hrt_abstime TRY_SUBSCRIBE_INTERVAL {1000 * 1000};
#endif

The second issue appears to be an increase in innovations due to incorrect time stamping of input data resulting in GPS data being retrieved from the buffer and at the wrong time, eg:

These are the original logged GPS velocity innovations for the simulated flight which included an acceleration to the North followed by a a decel and stop, followed by an RTL on reciprocal course back to the takeoff location:

![Screen Shot 2020-04-28 at 2 57 28 pm](https://user-images.githubusercontent.com/3596952/80448742-9ea99100-8960-11ea-8c6f-1b242824fd57.png)

Here is the north acceleration value:
![Screen Shot 2020-04-28 at 2 58 42 pm](https://user-images.githubusercontent.com/3596952/80448812-d0225c80-8960-11ea-83cb-5798a111a117.png)

Small velocity innovations like this are common and normally caused by timing offsets and/or uncompensated offsets from IMU to GPS antenna combined with angular rates.  The magnitude of velocity innovations in the SITL log would be consistent with GPs being approx 70msec less delayed than the 110msec that the EKF2_GPS_DELAY parameter specifies. 

Replaying the log shows a much larger and opposite sign innovation levels during accel/decel:

![Screen Shot 2020-04-28 at 3 10 24 pm](https://user-images.githubusercontent.com/3596952/80449412-70c54c00-8962-11ea-8f30-7b3ecee9e693.png)

The sign and amplitude of the North velocity innovation during the initial acceleration phase (before the filter 'corrects' the states) is consistent with an additional 200msec (1 GPS frame) delay, eg:

![Screen Shot 2020-04-28 at 3 30 06 pm](https://user-images.githubusercontent.com/3596952/80450585-2a252100-8965-11ea-9272-ce2e2fff7904.png)

What this doesn't establish is whether the timing error is present in the original SITL log or introduced by the replay. Plotting the GPS and EKF north velocity from the original SITL log shows the GPS to be delayed relative to the EKF by over 350 msec according to its own timestamp, eg:

![Screen Shot 2020-04-28 at 3 35 54 pm](https://user-images.githubusercontent.com/3596952/80451008-26de6500-8966-11ea-8139-8afdbcb1d23e.png)

This suggests that the issue is caused by the timestamping of GPS data when it is logged in SITL

**Comment by priseborough** [comment]
The next step is to determine if this is a jmavsim issue of if it also affects logs generated using gazebo

**Comment by priseborough** [comment]
Gazebo has a GPS delay rel to the time stamp of around 110msec

![Screen Shot 2020-04-30 at 6 50 39 pm](https://user-images.githubusercontent.com/3596952/80691473-a9f0ee00-8b13-11ea-85c0-6b0f2b2e6964.png)

Here is the same code and post  takeoff manoeuvre flown using jmavsim which shows a 200msec delay

/home/parallels/src/Firmware/build/px4_sitl_default/tmp/rootfs/log/2020-04-30/08_48_00.ulg

It is looking like this GPS time stamp problem is specific to jmavsim

**Comment by priseborough** [comment]
Replaying these generated gazebo and jmavsim files and comparing GPS velocity innovations with the original SITL log using master at commit 0c8dcf94bc256801a45475cdd7bc6490e9fb8af7 gave the following:

jmavim results generated using following velocity profile:
![Screen Shot 2020-05-02 at 11 20 59 am](https://user-images.githubusercontent.com/3596952/80851774-07e21a80-8c67-11ea-8a14-aa4bd0d31342.png)

GPS velocity innovation from original jmavsim log - these look high and are consistent with the observed excessive lag of the GPS relative to truth data:
![jmavsim_original](https://user-images.githubusercontent.com/3596952/80851962-1e3ca600-8c68-11ea-8fe4-865051744f9d.png)

GPS velocity innovations from replayed jmavsim log - these are higher than the original log and indicate additional GPS timing delay:
![jmavsim_replay](https://user-images.githubusercontent.com/3596952/80851966-2399f080-8c68-11ea-8b26-6ebac3862172.png)

Gazebo results generated using following velocity profile:
![Screen Shot 2020-05-02 at 11 20 59 am](https://user-images.githubusercontent.com/3596952/80851984-390f1a80-8c68-11ea-9119-175a6d9b7265.png)

GPS velocity innovations from original gazebo log - these look good and indicate the time stamping of the GPS data received by the EKF is correct:
![gazebo_original](https://user-images.githubusercontent.com/3596952/80851989-3f9d9200-8c68-11ea-9d04-a72e6b2db94a.png)

GPS velocity innovations from replayed gazebo log - these are higher than the original log and indicate additional timing delay has been introduced:
![gazebo_replay](https://user-images.githubusercontent.com/3596952/80852065-d2d6c780-8c68-11ea-821c-080d4afc1e01.png)

In summary there are two error sources:

- jmavsim has excessive GPS delay
- The processing chain from logging of SITL data to replay of the same data has introduced what appears to be a one sample (approx 200msec) time delay in the GPS data relative to the IMU data.

The next step is to obtain flight test data using the same branch which will indicate if the additoinal GPS delay seen during replay is still present. If not then the focus of the investigation will be on the time stamping of the GPS data during logging of SITL data.

**Comment by priseborough** [comment]
Replay of a flight log generated from an earlier commit with compatible uORB messages showed that the effect was still present.

**Comment by priseborough** [comment]
The time offset during replay also affects Baro data and is consistent with the GPS offset, eg see correlation between vertical innovations and climb rate that is only present in replay:

Original Log:
![Screen Shot 2020-05-05 at 2 58 50 pm](https://user-images.githubusercontent.com/3596952/81036077-0754b880-8ee1-11ea-9d04-540ac6e1ad64.png)

Replayed Log:
![Screen Shot 2020-05-05 at 2 59 20 pm](https://user-images.githubusercontent.com/3596952/81036087-0d4a9980-8ee1-11ea-89fc-c7b005ed3830.png)

Magnetometer data as evidenced by innovations during a yaw spin also shows evidence of an offset that is consistent with a time stamp offset between 100and 200 msec:

Original Log:
![Screen Shot 2020-05-05 at 3 04 10 pm](https://user-images.githubusercontent.com/3596952/81036249-b6918f80-8ee1-11ea-9b8d-670c3926f227.png)

Replayed Log:
![Screen Shot 2020-05-05 at 3 05 03 pm](https://user-images.githubusercontent.com/3596952/81036276-d1fc9a80-8ee1-11ea-8335-cf5908d2fd87.png)

Given the number of sensor types affected, a replay bug affecting the time stamping of IMU data is possible.

**Comment by priseborough** [comment]
@bkueng Are you aware of any changes to replay that could be causing the sensor timing offset?

**Comment by bkueng** [comment]
No. Is that in simulation only?

jmavsim gps delay is explicitly set to 200: https://github.com/PX4/jMAVSim/blob/master/src/me/drton/jmavsim/Simulator.java#L452. We can reduce that.

**Comment by priseborough** [comment]
It seems to be affecting replay of both SITL and flight logs. I will confirm the latter with some data gathered using latest master.

**Comment by priseborough** [comment]
Replay is showing a large time slip (accumulated IMU delta time not matching lapsed time)

![Screen Shot 2020-06-09 at 6 43 34 pm](https://user-images.githubusercontent.com/3596952/84126497-51513100-aa81-11ea-8c07-90077f2d6fee.png)

**Comment by priseborough** [comment]
This positive gradient indicates that when replay is running, not all IMU data is being received or processed by the ekf2

**Comment by priseborough** [comment]
Running replay with a printf in ekf2_main.cpp to monitor the time difference between subsequent IMU data, eg:

static int64_t last_imu_time = 0;
printf("imu_time_delta_us=%i\n",(int64_t)imu_sample_new.time_us - last_imu_time);
last_imu_time = (int64_t)imu_sample_new.time_us;

Shows that frequently data is skipped and sometimes duplicated, eg:
imu_time_delta_us=8000
imu_time_delta_us=4000
imu_time_delta_us=4000
imu_time_delta_us=4000
imu_time_delta_us=4000
imu_time_delta_us=8000
imu_time_delta_us=4000
imu_time_delta_us=4000
imu_time_delta_us=8000
imu_time_delta_us=4000
imu_time_delta_us=4000
imu_time_delta_us=4000
imu_time_delta_us=4000
imu_time_delta_us=4000
imu_time_delta_us=4000
imu_time_delta_us=4000
imu_time_delta_us=4000
imu_time_delta_us=8000
imu_time_delta_us=0
imu_time_delta_us=4000
imu_time_delta_us=4000
imu_time_delta_us=8000
imu_time_delta_us=4000

**Comment by priseborough** [comment]
What I have been able to determine so far is that within 

ReplayEkf2::handleTopicUpdate 

the ekf2_timestamps data  retrieved from the file using 

memcpy(&ekf2_timestamps, data, sub.orb_meta->o_size);

is missing frames. I checked the ekf2_timestamps  data published to the uORB during SITL operation and there are no frames missing.

**Comment by priseborough** [comment]
Converting the SITL generated log file to csv and inspecting shows that the one frame gaps in the ekf2_timestamp data are present in the file, eg:

<img width="221" alt="Screen Shot 2020-06-11 at 4 41 48 pm" src="https://user-images.githubusercontent.com/3596952/84353448-833dd100-ac02-11ea-978b-da60a2884939.png">

Investigation will now ficus on the logger

**Comment by priseborough** [comment]
A bench test on a F7 board (Pixhawk4) shows that there are no missing samples, however the baseline EKF update rate for this hardware is 200Hz, eg:

<img width="277" alt="Screen Shot 2020-06-18 at 9 57 06 am" src="https://user-images.githubusercontent.com/3596952/84962404-1bc7da00-b14a-11ea-8322-476d391bdc69.png">

**Comment by priseborough** [comment]
Closing this as it has been picked up here: https://github.com/PX4/Firmware/issues/15088

I have confirmed with flying and replay today that with current upstream master, the logging defect does not affect the F7 hardware i fly with.

**Comment by bkueng** [comment]
Thanks @priseborough for clarifying. I'll fix the underlying issue in SITL.

**Comment by vanurag** [comment]
Thanks @priseborough for looking into the issue extensively. I can confirm that as per #15088, just by polling on ekf2_timestamps for now, I could get much better results in SITL.
